rfcs/react-components/convergence/component-transitions-on-mount-or-unmount.md

```diff
@@ -0,0 +1,239 @@
+
+To determine the motion state of a component, a hook could be created as part of the @fluentui/react-utilities package. A preliminary implementation of this hook can be found here: [useMotionPresence](https://github.com/marcosmoura/fluentui/blob/feat/drawer-motion/packages/react-components/react-utilities/src/hooks/useMotionPresence.ts).
+
+#### What is it?
+
+A tracker hook, that monitors the state of animations and transitions for a particular element. A hook to track the state of animations/transitions. This hook does not directly create animations but instead synchronizes with CSS properties to determine the rendering status, visibility, entering, leaving, and ongoing animation of a component. If any CSS changes or properties are overridden, this hook will automatically adjust and stay synchronized.
+
+#### Usage
+
+State:
+
+```ts
+import * as React from 'react';
+import { getNativeElementProps, useMotionPresence, useMergedRefs } from '@fluentui/react-utilities';
+
+import type { SampleProps, SampleState } from './Sample.types';
+
+export const useSample_unstable = ({ open = false }: SampleProps, ref: React.Ref<HTMLElement>): SampleState => {
+  const { ref: componentRef, shouldRender, visible, entering, exiting, animating } = useMotionPresence(open);
+
+  return {
+    components: {
+      root: 'div',
+    },
+
+    root: getNativeElementProps('div', {
+      ref: useMergedRefs(ref, componentRef),
+    }),
+
+    shouldRender,
+    visible,
+    entering,
+    exiting,
+    animating,
+  };
+};
+```
+
+Renderization:
+
+```tsx
+import * as React from 'react';
+import { getSlots } from '@fluentui/react-utilities';
+import type { SampleState, SampleSlots } from './Sample.types';
+
+/**
+ * Render the final JSX of Sample
+ */
+export const renderSample_unstable = (state: SampleState) => {
+  const { slots, slotProps } = getSlots<SampleSlots>(state);
+
+  if (!state.shouldRender) {
+    return null;
+  }
+
+  return <slots.root {...slotProps.root} />;
+};
+```
+
+Styles:
+
+```ts
+import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
+import type { SampleSlots, SampleState } from './Sample.types';
+import type { SlotClassNames } from '@fluentui/react-utilities';
+import { tokens } from '@fluentui/react-theme';
+
+export const SampleClassNames: SlotClassNames<SampleSlots> = {
+  root: 'fui-Sample',
+};
+
+/**
+ * Styles for the root slot
+ */
+const useStyles = makeStyles({
+  root: {
+    opacity: 0,
+    transitionTimingFunction: 'ease-out',
+    transitionProperty: 'opacity',
+    willChange: 'opacity',
+  },
+
+  entering: {
+    transitionDuration: '200ms',
+  },
+
+  exiting: {
+    transitionDuration: '250ms',
+  },
+
+  visible: {
+    opacity: 1,
+  },
+});
+
+export const useSampleStyles_unstable = (state: SampleState): SampleState => {
+  const styles = useStyles();
+
+  state.root.className = mergeClasses(
+    SampleClassNames.root,
+    state.visible && styles.visible,
```

Comment:
Some kind of prop should be added so that the `visible` and `shouldRender` states are both `true` on initial mount.

Without that, any components that animate on mount will have a 'flash' because the component is mounted in DOM without animation styles, then the animation styles kick in on next render

Reply:
I can repro this on an experiment branch I made to test this with toast https://github.com/microsoft/fluentui/pull/28419

Reply:
This was fixed in the implementation branch. This document doesn't need any updates as this doesn't change the proposed API. :)